client/jvm/src/test/scala/org/http4s/client/scaffold/HandlersToNettyAdapter.scala

```diff
@@ -0,0 +1,158 @@
+import io.netty.channel._
+import io.netty.handler.codec.http.HttpHeaderNames._
+import io.netty.handler.codec.http.HttpResponseStatus._
+import io.netty.handler.codec.http.HttpVersion.HTTP_1_1
+import io.netty.handler.codec.http._
+import io.netty.util.CharsetUtil
+import org.log4s.getLogger
+
+import java.net.URI
+
+object HandlersToNettyAdapter {
+  val defaultFallbackHandler: Handler = (ctx: ChannelHandlerContext, _: HttpRequest) => {
+    HandlerHelpers.sendResponse(ctx, HttpResponseStatus.NOT_FOUND)
+    ()
+  }
+
+  def apply[F[_]](
+      handlers: Map[(HttpMethod, String), Handler],
+      fallbackHandler: Handler = defaultFallbackHandler,
+  )(implicit F: Sync[F]): F[ChannelInboundHandler] =
+    F.delay(new HandlersToNettyAdapter(handlers, fallbackHandler))
+}
+
+class HandlersToNettyAdapter private (
+    handlers: Map[(HttpMethod, String), Handler],
+    fallbackHandler: Handler,
+) extends SimpleChannelInboundHandler[HttpObject] {
+
+  private val logger = getLogger(this.getClass)
+
+  private var currentRequest: HttpRequest = null
+  private var currentHandler: Handler = null
+
+  override def channelRead0(ctx: ChannelHandlerContext, msg: HttpObject): Unit = {
+
+    msg match {
+      case request: HttpRequest =>
+        logger.trace(
+          s"Recieved [${request.method()}] [${request.uri()}] request from [${ctx.channel.remoteAddress()}]."
+        )
+        if (HttpUtil.is100ContinueExpected(request)) {
+          ctx.write(new DefaultFullHttpResponse(HTTP_1_1, CONTINUE))
+        } else {
+          currentRequest = request
+          currentHandler = handlers.getOrElse(
+            (request.method(), new URI(request.uri()).getPath()),
+            fallbackHandler,
+          )
+          currentHandler.onRequestStart(ctx, currentRequest)
+        }
+      case _ =>
+    }
+
+    msg match {
+      case content: HttpContent =>
+        logger.trace("Recieved content.")
+        currentHandler.onContent(ctx, currentRequest, content)
+      case _ =>
+    }
+
+    msg match {
+      case _: LastHttpContent =>
+        logger.trace("Request finished.")
+        currentHandler.onRequestEnd(ctx, currentRequest)
+        currentRequest = null
+        currentHandler = null
+      case _ =>
+    }
+  }
+
+  override def channelReadComplete(ctx: ChannelHandlerContext): Unit = {
+    ctx.flush()
+    ()
+  }
+
+  override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable): Unit = {
+    logger.warn(cause)("")
+    ctx.close()
+    ()
+  }
+
+}
+
+trait Handler {
+
+  def onRequestStart(
+      @deprecated("unused", "") ctx: ChannelHandlerContext,
+      @deprecated("unused", "") request: HttpRequest,
+  ): Unit = ()
+
+  def onContent(
+      @deprecated("unused", "") ctx: ChannelHandlerContext,
+      @deprecated("unused", "") request: HttpRequest,
+      @deprecated("unused", "") content: HttpContent,
+  ): Unit = ()
+
+  def onRequestEnd(ctx: ChannelHandlerContext, request: HttpRequest): Unit
```

Comment:
Is there any particular reason this one didn't have a default implementation, too?

Reply:
The idea was to make this trait Single-Abstract-Method so that in simple cases it can be implemented with the lambda expression syntax. The simple cases ended up being implemented with `RoutesToNettyAdapter` so didn't use this capability, but I still think it makes sense to have it like this. 

There is one place where I could actually use it. The '/process-request-entity'. But I opted for using the anonymous class syntax to stay consistent with `/respond-and-close-immediately` and `/respond-and-close-immediately-no-body`.